### ISSUE TITLE
Frontend doesn't emit IR (or sanitizer checks) for discarded-value expression (even when they contain UB)

### ISSUE THREAD

**Issue by Teemperor** [llvm/llvm-project#55382]
So it turns out Clang doesn't emit any IR for loads from pointers if the result of the load isn't used for anything.

A simple case is the example below:

```cpp
int main() {
    int *i = nullptr;
    (void)(*i); // This should be reported by UBSan?
}
```
On godbolt:
https://godbolt.org/z/veE1cWbe8

One way how this can occur in the wild is via the "assert with preserving side effects trick" such as:

```cpp
#define safe_assert(s) do { (s); } while(false)
```

The main problem here is that with enabled sanitizer this issue doesn't get caught or reported as there is just no code that can be sanitized. Note that once you change the expression from `*i` to like `*i + 0` then you actually get a correct sanitizer check emitted.

This actually seems to happen in the frontend which doesn't generate any IR (so, this is not DCE or anything like that).
